### ISSUE TITLE
Use go-fuzz for improved testing

### ISSUE THREAD

**Issue by Metalnem** [unidoc/unidoc#81]
Using [go-fuzz](https://github.com/dvyukov/go-fuzz) for around 30 minutes with just three small random PDF files, I was able to find multiple bugs in the PDF reader. You should probably fuzz it more and with improved corpus. For more details on fuzzing in Go, see [this](https://medium.com/@dgryski/go-fuzz-github-com-arolek-ase-3c74d5a3150c). My Fuzz function was looking like this:

```go
func Fuzz(data []byte) int {
	b := bytes.NewReader(data)

	if _, err := model.NewPdfReader(b); err != nil {
		return 0
	}

	return 1
}
```

**Comment by gunnsth** [comment]
We will definitely look into this.  We already have a pretty big test corpus and handle a lot of all kinds of files, although most of our focus has been on files that are incompatible with the PDF standard but can still be opened by PDF readers as they are very forgiving.  This will be a good chance to take our tests further and make more robust.

Many thanks.  I had not heard of go-fuzz before, but it looks perfect for this.

**Comment by gunnsth** [comment]
So far, beyond the 4 issues found and reported by @Metalnem initially:

https://github.com/unidoc/unidoc/issues/77
panic: interface conversion: pdf.PdfObject is nil, not *pdf.PdfObjectInteger bug
Fixed by commit 1769c68234b3544230dc5e2fc38f75eb42b54d94

https://github.com/unidoc/unidoc/issues/78
panic: runtime error: makeslice: len out of range
Fixed by commit 68bca163a9d1d3f46e76f0ecab9cbf7d67f0f633

https://github.com/unidoc/unidoc/issues/79
panic: runtime error: invalid memory address or nil pointer dereference bug
Fixed by 5f506b74274b9e4f44f0d4c2ad4803bcdb4ed910

https://github.com/unidoc/unidoc/issues/80
runtime: goroutine stack exceeds 1000000000-byte limit bug
Fixed by commit 6e30b10029c2676f49ec5a850185ea59c7736a92

We have fuzzed, found and fixed additional 8 bugs (so total of 12 reported & fixed):

https://github.com/unidoc/unidoc/commit/469856432bb46537f8ded7752d29becec3cef8a2
Fix fuzz-detected problem where stream length referring to stream itself (with a test case to verify)

https://github.com/unidoc/unidoc/commit/b2bb6d983607309417c068b6f38448b7b1bf8c63
Fixed problem found by fuzzing when Encrypt pointing to invalid reference

https://github.com/unidoc/unidoc/commit/99872f29ba531010e26580530a142d8e3335b9b4
Type check for xref Prev entries added.

https://github.com/unidoc/unidoc/commit/124bd10e7cbc848c63a0d41d76328c77e21f29f4
Check bounds on XRef stream index list (found by fuzzing)

https://github.com/unidoc/unidoc/commit/3639b5c2794d356ceb0709ed0d2a81c64fef287b
Add check for ID array length

https://github.com/unidoc/unidoc/commit/7f83a4ea2fd23f5321254a758ebd1286313b940e
Add bounds check in crypt, parser. Avoid division by zero.  (3 fixes)

We have created a larger and more optimized test corpus and are in the process of running it on a pretty powerful machine for a few days.  We will create an issue for each of the bugs found and refer to it here.

Our goal with this work is to make unidoc as stable as possible. Bugs like these can potentially be used for denial of service attacks.  Fortunately, they cannot be used for remote code execution attacks.
